skbio/metadata/_interval.py

```diff
@@ -0,0 +1,699 @@
+    Interval(interval_metadata=..., locations=[(3, 9)], \
+boundaries=[(True, True)], metadata={'gene': 'sagB'})
+    >>> im.add(locations=[(3, 7)],
+    ...        metadata={'gene': 'sagC'})  # doctest: +ELLIPSIS
+    Interval(interval_metadata=..., locations=[(3, 7)], \
+boundaries=[(True, True)], metadata={'gene': 'sagC'})
+    >>> im.add(locations=[(1, 2), (4, 7)],
+    ...        metadata={'gene': 'sagA'})  # doctest: +ELLIPSIS
+    Interval(interval_metadata=..., locations=[(1, 2), (4, 7)], \
+boundaries=[(True, True), (True, True)], metadata={'gene': 'sagA'})
+
+    Show the object representation:
+
+    >>> im    # doctest: +ELLIPSIS
+    3 interval features
+    -------------------
+    Interval(interval_metadata=..., locations=[(3, 9)], \
+boundaries=[(True, True)], metadata={'gene': 'sagB'})
+    Interval(interval_metadata=..., locations=[(3, 7)], \
+boundaries=[(True, True)], metadata={'gene': 'sagC'})
+    Interval(interval_metadata=..., locations=[(1, 2), (4, 7)], \
+boundaries=[(True, True), (True, True)], metadata={'gene': 'sagA'})
+
+    We can sort the genes by their locations:
+
+    >>> im.sort()
+    >>> im    # doctest: +ELLIPSIS
+    3 interval features
+    -------------------
+    Interval(interval_metadata=..., locations=[(1, 2), (4, 7)], \
+boundaries=[(True, True), (True, True)], metadata={'gene': 'sagA'})
+    Interval(interval_metadata=..., locations=[(3, 7)], \
+boundaries=[(True, True)], metadata={'gene': 'sagC'})
+    Interval(interval_metadata=..., locations=[(3, 9)], \
+boundaries=[(True, True)], metadata={'gene': 'sagB'})
+
+    Query the genes by location and/or metadata:
+
+    >>> q1 = im.query([(7, 9)])
+    >>> list(q1)  # doctest: +ELLIPSIS
+    [Interval(interval_metadata=..., locations=[(3, 9)], \
+boundaries=[(True, True)], metadata={'gene': 'sagB'})]
+    >>> q2 = im.query(metadata={'gene': 'sagA'})
+    >>> list(q2)  # doctest: +ELLIPSIS
+    [Interval(interval_metadata=..., locations=[(1, 2), (4, 7)], \
+boundaries=[(True, True), (True, True)], metadata={'gene': 'sagA'})]
+    >>> q3 = im.query([(1, 2)], metadata={'gene': 'foo'})
+    >>> list(q3)
+    []
+
+    Drop genes:
+
+    >>> im.drop(metadata={'gene': 'sagA'})
+    >>> im.sort()
+    >>> im   # doctest: +ELLIPSIS
+    2 interval features
+    -------------------
+    Interval(interval_metadata=..., locations=[(3, 7)], \
+boundaries=[(True, True)], metadata={'gene': 'sagC'})
+    Interval(interval_metadata=..., locations=[(3, 9)], \
+boundaries=[(True, True)], metadata={'gene': 'sagB'})
+    """
+    def __init__(self):
+        # List of Interval objects.
+        self._intervals = []
+
+        # IntervalTree object to allow faster querying of interval objects.
+        self._interval_tree = IntervalTree()
+
+        # Indicates if the IntervalTree needs to be rebuilt.
+        self._is_stale_tree = False
+
+    def _rebuild_tree(method):
+        """Rebuild the IntervalTree."""
+        @wraps(method)
```

Comment:
alternatively you could use this decorator only on private methods, where the public signature doesn't matter at all (as far as I can tell this would only impact `query` resulting in a need for `_query`, but that might be preferable to you). Either way of dealing with this is fine with me.